base/compiler/abstractinterpretation.jl

```diff
@@ -2243,6 +2243,12 @@ function abstract_eval_statement_expr(interp::AbstractInterpreter, e::Expr, vtyp
             merge_effects!(interp, sv, effects)
             if isa(sv, InferenceState)
                 sv.stmt_info[sv.currpc] = info
+                # mark this call statement as DCE-elgible
+                if is_removable_if_unused(effects)
+                    add_curr_ssaflag!(sv, IR_FLAG_EFFECT_FREE)
+                else
+                    sub_curr_ssaflag!(sv, IR_FLAG_EFFECT_FREE)
```

Comment:
Would it be better to do this in a single pass at the end based on the `info` object rather than toggling the flag back and forth here?

Reply:
Hm, yeah. The (unmentioned) support for `:foreigncall` isn't that important anyway.

Reply:
Well, we need something like #46962 (which came with a minor computational cost when I tried last time).

Reply:
Since I couldn't fix the performance regression of #46962 , I'd like to merge this as is leaving this as a TODO comment.